File: 202-happy-number/202-happy-number.cpp
class Solution {
public:
    bool isHappy(int n) {
        unordered_map<int, int> visited;       
        
        while (true) {
            if (n == 1) 
                return true;
            
            if (visited[n])
                return false;
            
            visited[n]++;
            int step1 = 0;
            
            while(n) {
                int temp = n % 10;
                temp *= temp;
                step1 += temp;
                n /= 10;
            }
            n = step1;
        }        
        return true;
    }
};

File: 202-happy-number/202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        
        visited = set()
        
        if n == 1:
            return True
        
        while (n != 1 and n not in visited):
            visited.add(n)
            digits = str(n)
            n = 0
            
            for i in digits:
                n += int(i)**2
            
            if n == 1:
                return True
        
        return False